src/bundler.ts

```diff
@@ -118,62 +116,138 @@ class Bundler {
+   *     for hidden div adjacency etc.
+   */
+  async inlineHtmlImport(
+      documentUrl: string, htmlImport: ASTNode,
       inlined: Set<string>): Promise<void> {
-    const rawUrl: string = dom5.getAttribute(htmlImport, 'href');
+    const rawUrl: string = dom5.getAttribute(htmlImport, 'href')!;
     const resolved = url.resolve(documentUrl, rawUrl);
     if (inlined.has(resolved)) {
       dom5.remove(htmlImport);
       return;
     }
     inlined.add(resolved);
     const backingDocument: ScannedDocument =
-        await analyzer._analyzeResolved(resolved);
+        await this.analyzer._scanResolved(resolved);
     const documentAst = dom5.parseFragment(backingDocument.document.contents);
     this.pathResolver.resolvePaths(documentAst, resolved, documentUrl);
-    const importParent = htmlImport.parentNode;
+    let importParent: ASTNode;
+    if (matchers.afterHiddenDiv(htmlImport)) {
+      importParent = dom5.nodeWalkPrior(htmlImport, matchers.hiddenDiv)!;
+      dom5.remove(htmlImport);
+      dom5.append(importParent, htmlImport);
+    } else if (matchers.beforeHiddenDiv(htmlImport)) {
+      const index = htmlImport.parentNode!.childNodes!.indexOf(htmlImport);
+      importParent = htmlImport.parentNode!.childNodes![index + 1];
+      dom5.remove(htmlImport);
+      ASTUtils.prepend(importParent, htmlImport);
+    } else if (!matchers.inHiddenDiv(htmlImport)) {
+      const hiddenDiv = this.getHiddenNode();
+      dom5.replace(htmlImport, hiddenDiv);
+      dom5.append(hiddenDiv, htmlImport);
+      importParent = hiddenDiv;
+    } else {
+      importParent = htmlImport.parentNode!;
+    }
+
+    ASTUtils.insertAllBefore(importParent, htmlImport, documentAst.childNodes!);
     dom5.remove(htmlImport);
-    ASTUtils.prependMultiple(importParent, documentAst.childNodes);
   }
 
+  /**
+   * Given a URL to an entry-point html document, produce a single document
+   * suitable for deployment.
+   */
   async bundle(url: string): Promise<ASTNode> {
-    const analyzer: Analyzer = new Analyzer(this.opts);
-    const analyzed: Document = await analyzer.analyzeRoot(url);
-    const newDocument = dom5.parse(analyzed.parsedDocument.contents);
-    const body = dom5.query(newDocument, matchers.body);
-    const head = dom5.query(newDocument, matchers.head);
-    const getNextImport = () => dom5.query(newDocument, matchers.htmlImport);
+    const analyzedRoot = await this.analyzer.analyzeRoot(url);
+    // TODO(usergenic): Don't re-parse this document if you already have an AST
+    // in parsedDocument.  Try cloning or even building newDocument as analyzed
+    // document is walked.
+    const newDocument = dom5.parse(analyzedRoot.parsedDocument.contents);
+
+    // Create a hidden div to target.
+    const body: ASTNode = dom5.query(newDocument, matchers.body)!;
+    const hiddenDiv = this.getHiddenNode();
+
+    const getNextHtmlImport = () =>
+        dom5.query(newDocument, matchers.htmlImport);
```

Comment:
this is going to query each time its called. Would it be correct to queryAll once and iterate through the list?